src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -349,14 +398,69 @@ fn compute_deps(
     Ok(ret)
 }
 
+fn package_lib(dep_pkg: &Package, has_artifact: bool, artifact_lib: bool) -> Option<&Target> {
+    dep_pkg.targets().iter().find(|t| {
+        if has_artifact {
+            t.is_lib() && artifact_lib
+        } else {
+            t.is_lib()
+        }
+    })
+}
+
+/// Adds build requests for artifact dependencies of `unit`.
+fn calc_artifact_deps(
+    unit: &Unit,
+    unit_for: UnitFor,
+    dep_id: PackageId,
+    deps: &HashSet<Dependency>,
+    state: &State<'_, '_>,
+    filter: &dyn Fn(&Unit, &Dependency) -> bool,
+    ret: &mut Vec<UnitDep>,
+) -> CargoResult<(bool, bool)> {
+    let mut has_artifact = false;
+    let mut artifact_lib = false;
+    let artifact_pkg = state.get(dep_id);
+    for (dep, artifact) in deps
+        .iter()
+        .filter(|dep| filter(unit, dep))
+        .filter_map(|dep| dep.artifact().map(|a| (dep, a)))
```

Comment:
This test was added specifically in 61b285cdd4a2ff45a5d830b3a95f6496f9fa4601 but worked even with the `filter_map()`. I think it will be good to have though as I start to untangle the filtering.

Reply:
Maybe the reason for the `normal_build_deps_are_picked_up_in_presence_of_an_artifact_build_dep_to_the_same_package()` test to work against expectations is that build dependencies are handled so differently.

Maybe there is a way to cause an issue if build dependencies wouldn't have to be handled specially as they are now (see [this comment](https://github.com/rust-lang/cargo/pull/9992#discussion_r808883418) for a little more detail.

As it stands, unfortunately I can't do anything about it :/.